lib/application/crowdaction/subscription/subscription_bloc.dart

```diff
@@ -60,9 +91,128 @@ class SubscriptionBloc extends Bloc<SubscriptionEvent, SubscriptionState> {
 
     emit(
       failureOrSuccess.fold(
-        (failure) => const SubscriptionState.subscriptionFailed(),
-        (_) => const SubscriptionState.subscribed(),
+        (failure) => SubscriptionState.subscriptionFailed(
+          commitments: state.commitments,
+          activeCommitments: state.activeCommitments,
+          deactivated: state.deactivated,
+        ),
+        (_) => SubscriptionState.subscribed(
+          commitments: state.commitments,
+          activeCommitments: state.activeCommitments,
+          deactivated: state.deactivated,
+        ),
+      ),
+    );
+  }
+
+  FutureOr<void> _mapSetActiveCommitmentsToState(
+    _SetActiveCommitment value,
+    Emitter<SubscriptionState> emit,
+  ) {
+    emit(state.copyWith(activeCommitments: value.commitments));
+  }
+
+  FutureOr<void> _mapSelectCommitmentToState(
+    _SelectCommitment value,
+    Emitter<SubscriptionState> emit,
+  ) {
+    _selectCommitment(value.commitmentOption);
+    emit(
+      state.copyWith(
+        activeCommitments: List.of(_activeCommitments),
+      ),
+    );
+  }
+
+  FutureOr<void> _mapDeselectCommitmentToState(
+    _DeselectCommitment value,
+    Emitter<SubscriptionState> emit,
+  ) {
+    _deselectCommitment(value.commitmentOption);
+    emit(
+      state.copyWith(
+        activeCommitments: List.of(_activeCommitments),
       ),
     );
   }
+
+  FutureOr<void> _mapResetToState(
+    _Reset value,
+    Emitter<SubscriptionState> emit,
+  ) {
+    emit(SubscriptionState.initial());
+    _commitments.clear();
+    _activeCommitments.clear();
+  }
+
+  FutureOr<void> _mapSetCommitmentsToState(
+    _SetCommitments value,
+    Emitter<SubscriptionState> emit,
+  ) {
+    emit(state.copyWith(commitments: value.crowdAction.commitmentOptions));
+    _commitments.clear();
+    _commitments.addAll(value.crowdAction.commitmentOptions);
+  }
+
+  FutureOr<void> _mapSetDeactivatedToState(
+    _SetDeactivatedState value,
+    Emitter<SubscriptionState> emit,
+  ) {
+    emit(state.copyWith(deactivated: value.isDeactivated));
+  }
+
+  final List<CommitmentOption> _commitments = [];
+  List<String> _activeCommitments = [];
+
+  void _selectAll(List<CommitmentOption> commitments) {
+    for (final commitment in _commitments) {
+      if (commitments.contains(commitment)) {
+        _activeCommitments.add(commitment.id);
+        _activeCommitments = _activeCommitments.toSet().toList();
+
+        final children = commitment.requires;
+        if (children != null) {
+          _selectAll(children);
+        }
+      }
+    }
+  }
```

Comment:
```suggestion
  void _selectAll() {
    for (final commitment in _commitments) {
      if (!_activeCommitments.contains(commitment.id)) {
        _activeCommitments.add(commitment.id);
      }
    }
  }
```
Does `_commitments` contain all the possible commitments? Because if yes, this doesn't need to be a recursive function, we can just iterate that list once and don't have to go through the `children`?

I see that it iterates `_commitments` and not the argument `commitments`, that's why I'm a bit confused.

Reply:
If we need it, I think we shouldn't do recursive function, recursive functions aren't always super easy to read. This could do it instead:
```suggestion
  void _selectAll(List<CommitmentOption> commitments) {
    List<CommitmentOption>? children;
    do {
      for (final commitment in commitments) {
        _activeCommitments.add(commitment.id);
        _activeCommitments = _activeCommitments.toSet().toList();

        children = commitment.requires;
      }
    } while (children != null);
  }
```